From ba292ba4e92b8e4c83de1d297b9082323cabfbf4 Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Mon, 9 Jun 2025 01:53:34 +0200
Subject: [PATCH] kbuild: support x86_64 ISA levels

This is a slightly reworked version of graysky's lite patch with extra
safety guards from Clear Linux applied.

Link: https://github.com/graysky2/kernel_compiler_patch
Link: https://github.com/clearlinux-pkgs/linux/blob/main/0133-novector.patch
Link: https://lore.kernel.org/lkml/20241210144945.2325330-4-arnd@kernel.org/
Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 arch/x86/Kconfig.cpu | 9 +++++++++
 arch/x86/Makefile    | 7 ++++++-
 2 files changed, 15 insertions(+), 1 deletion(-)

diff --git a/arch/x86/Kconfig.cpu b/arch/x86/Kconfig.cpu
index f928cf6e3..7f2bc6a30 100644
--- a/arch/x86/Kconfig.cpu
+++ b/arch/x86/Kconfig.cpu
@@ -281,6 +281,15 @@ config X86_GENERIC
 	  This is really intended for distributors who need more
 	  generic optimizations.
 
+config X86_64_ISA_LEVEL
+	int "x86_64 ISA level"
+	range 1 4
+	depends on (CC_IS_GCC && GCC_VERSION > 110000) || (CC_IS_CLANG && CLANG_VERSION >= 120000)
+	depends on X86_64
+	depends on !X86_NATIVE_CPU
+	help
+	  Specify an x86_64 ISA level.
+
 #
 # Define implied options from the CPU selection here
 config X86_INTERNODE_CACHE_SHIFT
diff --git a/arch/x86/Makefile b/arch/x86/Makefile
index 4db7e4bf6..9f480f15b 100644
--- a/arch/x86/Makefile
+++ b/arch/x86/Makefile
@@ -75,7 +75,7 @@ export BITS
 #
 #    https://gcc.gnu.org/bugzilla/show_bug.cgi?id=53383
 #
-KBUILD_CFLAGS += -mno-sse -mno-mmx -mno-sse2 -mno-3dnow -mno-avx
+KBUILD_CFLAGS += -mno-sse -mno-mmx -mno-sse2 -mno-3dnow -mno-avx -mno-avx2 -fno-tree-vectorize
 KBUILD_RUSTFLAGS += --target=$(objtree)/scripts/target.json
 KBUILD_RUSTFLAGS += -Ctarget-feature=-sse,-sse2,-sse3,-ssse3,-sse4.1,-sse4.2,-avx,-avx2
 
@@ -166,8 +166,13 @@ ifdef CONFIG_X86_NATIVE_CPU
         KBUILD_CFLAGS += -march=native
         KBUILD_RUSTFLAGS += -Ctarget-cpu=native
 else
+    ifeq ($(CONFIG_X86_64_ISA_LEVEL),1)
         KBUILD_CFLAGS += -march=x86-64 -mtune=generic
         KBUILD_RUSTFLAGS += -Ctarget-cpu=x86-64 -Ztune-cpu=generic
+    else
+	KBUILD_CFLAGS += -march=x86-64-v$(CONFIG_X86_64_ISA_LEVEL)
+	KBUILD_RUSTFLAGS += -Ctarget-cpu=x86-64-v$(CONFIG_X86_64_ISA_LEVEL)
+    endif
 endif
 
         KBUILD_CFLAGS += -mno-red-zone
-- 
2.51.0

